src/lightmap.cpp

```diff
@@ -469,6 +470,173 @@ bool map::pl_sees( const tripoint &t, const int max_range ) const
           map_cache.sm[t.x][t.y] > 0.0 );
 }
 
+template<int xx, int xy, int xz, int yx, int yy, int yz, int zz,
+         float(*calc)(const float &, const float &, const int &),
+         bool(*check)(const float &, const float &)>
+void cast_zlight(
+    float (&output_cache)[MAPSIZE*SEEX][MAPSIZE*SEEY],
+    const std::array<const float (*)[MAPSIZE*SEEX][MAPSIZE*SEEY], OVERMAP_LAYERS> &input_arrays,
+    const tripoint &offset, const int offset_distance, const int target_z,
+    const std::function<bool(const tripoint &)> &floor_check,
```

Comment:
Big red flag here (floor_check), the reason I supplied calc and check as template arguments is that dispatching to std::function objects is very expensive in a loop this tight.  I had no idea it was as bad as it is before I tested it, but it's BAAD.